alf/networks/relu_mlp.py

```diff
@@ -179,3 +178,45 @@ def _compute_jac_diag(self):
                              self._fc_layers[0].weight)  # [B, n]
 
         return J
+
+    def compute_vjp(self, inputs, vec):
+        """Compute vector-Jacobian product. 
+
+        Args:
+            inputs (Tensor): size (self._input_size) or (B, self._input_size)
+            vec (Tensor): the vector for which the vector-Jacobian product
+                is computed. Must be of size (self._output_size) or
+                (B, self._output_size). 
+
+        Returns:
+            vjp (Tensor): size (self._input_size) or (B, self._input_size).
+        """
+
+        assert inputs.ndim == vec.ndim, \
+            ("ndim of inputs and vec must be consistent!")
+        if inputs.ndim > 1:
+            assert inputs.ndim == 2, \
+                ("inputs must be a vector or matrix!")
+            assert inputs.shape[0] == vec.shape[0], \
+                ("batch size of inputs and vec must agree!")
+        assert inputs.shape[-1] == self._input_size, \
+            ("inputs should has shape {}!".format(self._input_size))
+        assert vec.shape[-1] == self._output_size, \
+            ("vec should has shape {}!".format(self._output_size))
+
+        self.forward(inputs)
+
+        return self._compute_vjp(vec)
+
+    def _compute_vjp(self, vec):
+        """Compute vector-Jacobian product. """
+
+        if vec.ndim == 1:
+            vec = vec.unsqueeze(0)
+
+        J = torch.matmul(vec, self._fc_layers[-1].weight)
+        for fc in reversed(self._fc_layers[0:-1]):
+            mask = (fc.hidden_neurons > 0).float()
+            J = torch.einsum('ba, ba, aj->bj', J, mask, fc.weight)
```

Comment:
I don't have very high confidence on the performance of einsum. matmul is generally highly optimized. And hence ``matmul(J * mask, fc.weight)`` might be faster.

Reply:
updated.